runtimebp/cpu.go

```diff
@@ -127,11 +131,83 @@ func defaultMaxProcsFormula(n float64) int {
 	return int(math.Ceil(n))
 }
 
-// GOMAXPROCS sets runtime.GOMAXPROCS with the default formula,
+func scaledMaxProcsFormula(n float64) int {
+	i := fetchMaxProcsMult()
+	return int(math.Ceil(n * float64(i)))
+}
+
+func fetchMaxProcsMult() int {
+	// Allow using a multiplier for number of processes relative to limit.
+	// Not catching the ok here because this function will never be called unless
+	// the environment variable is set
+	v, _ := os.LookupEnv("GOMAXPROCSMULT")
+	i, err := strconv.Atoi(v)
+	if err != nil {
+		fmt.Fprintf(
+			os.Stderr,
+			"runtimebp.fetchMaxProcsMult: GOMAXPROCSMULT is set to %q, which is invalid, falling back to 2 to ensure parallelism: %v",
+			v,
+			err,
+		)
+		i = 2
+	}
+	if i <= 0 {
+		i = 1
+	}
+	return i
+}
+
+// fetchCpuRequest checks for the magic CPUREQUEST variable set as a fallback
+// by Infrared.  If this is set, we use it and the multiplier to set our
+// GOMAXPROCS as a fallback.
+func fetchCPURequest() (int, error) {
+	if v, ok := os.LookupEnv("CPUREQUEST"); ok {
+		req, err := strconv.Atoi(v)
+		if err != nil {
+			// This case will probably never be hit, but I'm keeping it just in case.
+			// In k8s 1.21 and newer, when you present the CPU request via the Downward API
+			// k8s automatically rounds it up to the nearest integer unit.  That's how
+			// we plan to set this variable, but just in case it might be worth doing this
+			// check in case they ever break that.  There's no official contract for that.
+			match := millicoreRegexp.FindStringSubmatch(v)
+			if match == nil {
+				fmt.Fprintf(
+					os.Stderr,
+					"runtimebp.fetchCPURequest: CPUREQUEST is set to %q, which is invalid, ignoring.",
+					v,
+				)
+				req = 1
+			} else {
+				m, err := strconv.Atoi(match[0])
+				if err != nil {
+					req = int(math.Ceil(float64(m) / 1000.0))
+				} else {
+					fmt.Fprintf(
```

Comment:
:no_bell: based on the regexp I don't think this case is possible but ¯\\\_(ツ)\_/¯

Reply:
I think you're actually right, and if you want I can remove it.

Reply:
we probably still want the error handling branch, even though the error is theoretically impossible to happen. we have some "prior arts" of that by just adding a comment of `// should not happen, but just in case` and still do error logging/whatever is appropriate